lib/compat/wordpress-6.6/resolve-patterns.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/**
+ * Replaces pattern blocks with their content.
+ *
+ * @param array $blocks Array of blocks.
+ * @param array $inner_content Optional array of inner content.
+ * @return array Array of blocks with patterns replaced.
+ */
+function gutenberg_replace_pattern_blocks( $blocks, &$inner_content = null ) {
+	// Keep track of seen references to avoid infinite loops.
+	static $seen_refs = array();
+	$i                = 0;
+	while ( $i < count( $blocks ) ) {
+		if ( 'core/pattern' === $blocks[ $i ]['blockName'] ) {
+			$slug = $blocks[ $i ]['attrs']['slug'];
+
+			if ( isset( $seen_refs[ $slug ] ) ) {
+				// Skip recursive patterns.
+				array_splice( $blocks, $i, 1 );
```

Comment:
A downside (but not a blocker) of this simple skipping is that recursion violations become silent in the editor, reducing users' ability to spot and fix them, but they will still be flagged in the front end by a different mechanism. 

In a production environment (with `false == WP_DEBUG && WP_DEBUG_DISPLAY`), that's fine, but in a development environment it could cause situations where a user sees a notice (`[block rendering halted]`) in the front end but no signs of recursion in the editor. At least theoretically — I haven't tried this.

Reply:
> Anyway, I think we could explore this separately if needed. I guess the main thing is preventing an infinite loop. :)

Like I said, not a blocker. :) 

> Yes, but we don't have mechanism to "store" errors in block markup. You just injected some text, but that doesn't work in the editor. We can't just inject a paragraph block that is then editable. 🤔 Maybe set an attribute and error in the editor if that's present? 

Could be a simple as a special block, no? Inject: 

```php
[
    'blockName'    => 'core/error',
    'attrs'        => [
        'type'    => 'recursion',
        'details' => $pattern_slug,
    ],
    'innerBlocks'  => [],
    'innerHTML'    => '',
    'innerContent' => [],
];